main/src/util/profiler.hpp

```diff
@@ -0,0 +1,147 @@
+    std::vector<float>  g1PerStep;     // skewness
+    std::vector<float>  g2PerStep;     // kurtosis
+    std::vector<float>  totalTimeStep; // total time-step durations
+    std::vector<size_t> numLocalParts; // number of local particles
+    std::vector<float>  funcTimes;     // vector of timesteps for all ranks
+    int                 _numRanks;
+    int                 _rank;
+    std::ofstream       profilingFile;
+
+public:
+    Profiler(int rank)
+        : _rank(rank)
+    {
+        MPI_Comm_size(MPI_COMM_WORLD, &_numRanks);
+        std::string filename = "profiling-" + std::to_string(_numRanks) + "Ranks.txt";
+        if (_rank == 0) profilingFile.open(filename);
+    }
+    ~Profiler() {}
+
+    void printMetrics(int iteration)
+    {
+        profilingFile << meanPerStep.at(iteration) << "," << stdevPerStep.at(iteration) << ","
+                      << covPerStep.at(iteration) << "," << I_2PerStep.at(iteration) << ","
+                      << lambdaPerStep.at(iteration);
+        for (int i = 0; i < _numRanks; i++)
+        {
+            profilingFile << "," << vectorMetric.at(iteration)[i];
+        }
+    }
+
+    void printProfilingInfo()
+    {
+        if (_rank == 0)
+        {
+            size_t iter = 1;
+            // profilingFile << std::setw(3) << " ";
+            for (int i = 0; i < _numRanks; i++)
+            {
+                profilingFile << "RANK" << i << ",";
+            }
+            profilingFile << "mean,"
+                          << "stdev,"
+                          << "CoV,"
+                          << "I_2,"
+                          << "lambda";
+            for (int i = 0; i < _numRanks; i++)
+            {
+                profilingFile << ",Vector" << i;
+            }
+            profilingFile << std::endl;
+            for (auto& element : timeSteps)
+            {
+                calculateMetrics(element);
+                // profilingFile << std::setw(3) << iter << " ";
+                for (int i = 0; i < _numRanks; i++)
+                {
+                    profilingFile << element[i] << ",";
+                }
+                printMetrics(iter - 1);
+                iter++;
+                profilingFile << std::endl;
+            }
+            profilingFile << std::endl;
+            std::cout << "Profiling data written." << std::endl;
+        }
+        profilingFile.close();
+    }
+
+    void saveTimings(float duration) { funcTimes.push_back(duration); }
+
+    void saveTimestepAndParticleNo(float duration, size_t numLocalParticles)
+    {
+        funcTimes.push_back(duration);                 // save the total timestep timing
+        funcTimes.push_back((float)numLocalParticles); // save the local number of particles
+    }
+
+    void gatherTimings()
+    {
+        std::cout << "Gathering profiling data." << std::endl;
+
+        float* durations = new float[_numRanks * funcTimes.size()];
```

Comment:
Memory leak here. Why use error-prone `new` here instead of an `std::vector` whose buffer you can access with `data()`?